Implement new exercise tree endpoint. Fixes #603

skulonen opened the pull request:
# Description

**What?**

Add new exercise tree endpoint (`/api/v2/courses/{course_id}/tree`) to A+ API. The new endpoint presents the learning objects of a course in a tree-like hierarchy, unlike the regular `exercises` endpoint, which presents them as a flat list.

**Why?**
The flat exercise list does not sort the exercises in the hierarchical order. This new endpoint is also necessary for the IntelliJ plugin.

**How?**

Created a new viewset (`CourseExerciseTreeViewSet`) which accesses the cached content of the course instance, and two new serializers for outputting the data (`TreeCourseModuleSerializer`, `TreeExerciseSerializer`). The viewset doesn't derive from `ReadOnlyModelViewSet` and the serializers don't inherit from `AplusModelSerializer` because they don't operate on model objects. Instead, they operate on `dict`s that the course instance's cached content contains.

The reason for using the cached content is that it is already in a hierarchical structure.

Fixes #603


# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)


**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

Tested that the exercise tree was returned correctly in cases where there are multiple modules, where the hierarchy is deeply nested, where a module has no exercises, and other cases.

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [ ] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

skulonen commented:
I asked Olli Kinnunen for feedback regarding the structure of the response, and after discussing it with his team, Olli confirmed that it is satisfactory.

Mankro commented on course/api/full_serializers.py:
```diff
@@ -219,3 +223,16 @@ class Meta(CourseStudentGroupBriefSerializer.Meta):
 
+
+class TreeCourseModuleSerializer(serializers.Serializer):
+    id = serializers.IntegerField()
```
I wonder if we should include other additional `CourseModule` fields in this serializer.

skulonen replied:
The IntelliJ team was happy with the fields currently available in the response. But if you think some specific fields should be added, I can do that.

Mankro commented on course/api/views.py:
```diff
@@ -1,4 +1,4 @@
-from rest_framework import filters, generics, permissions, viewsets, status, mixins
+from rest_framework import filters, generics, permissions, viewsets, status, mixins, views
```
You added import for `views`, but is it actually used anywhere?

skulonen replied:
It is not, I'll remove it.

Mankro commented on course/api/views.py:
```diff
@@ -68,6 +68,20 @@ def get_object(self):
+
+    def list(self, request, *args, **kwargs):
+        serializer = self.serializer_class(
+            self.content.data['modules'],
```
Just writing down that `self.content` comes from the super class `CourseResouceMixin - CourseInstanceBaseMixin`. It is the course cache `CachedContent`.

* https://github.com/apluslms/a-plus/blob/46386ac304ebfb8dc00bfd2bffdb7d0d01d6ae8a/course/viewbase.py#L61
* https://github.com/apluslms/a-plus/blob/46386ac304ebfb8dc00bfd2bffdb7d0d01d6ae8a/exercise/cache/content.py#L10

Mankro reviewed (approved):
This looks excellent in a quick overview!

* Does the tree include both exercises and chapters? Looks like yes, just double-checking.
* You could add doc strings to the new classes. It is hard to see from the code that the data comes from the cache. The new serializer classes are very dependent on the structure of the cache object.

skulonen commented:
@Mankro I've added the doctrings, removed the unnecessary import and squashed the commits.

>Does the tree include both exercises and chapters? Looks like yes, just double-checking.

Yes. It has to, because exercise can be children of chapters, and things would become really messy if we tried to exclude the chapters but include their child exercises.
